Plugins/SwiftProtobufPlugin/plugin.swift

```diff
@@ -0,0 +1,139 @@
+import Foundation
+import PackagePlugin
+
+@main
+struct SwiftProtobufPlugin: BuildToolPlugin {
+    /// Errors thrown by the `SwiftProtobufPlugin`
+    enum PluginError: Error {
+        /// Indicates that the target where the plugin was applied to was not `SourceModuleTarget`.
+        case invalidTarget
+        /// Indicates that no config file was present in the root of the target directory.
+        case noConfigFile
+    }
+
+    /// The configuration of the plugin.
+    struct Configuration: Codable {
+        /// Encapsulates a single invocation of protoc.
+        struct Invocation: Codable {
+            /// The visibility of the generated files.
+            enum Visibility: String, Codable {
+                /// The generated files should have `internal` access level.
+                case `internal`
+                /// The generated files should have `public` access level.
+                case `public`
+            }
+
+            /// An array of paths to `.proto` files for this invocation.
+            var protoFiles: [String]
+            /// The visibility of the generated files.
+            var visibility: Visibility?
+        }
+
+        /// The path to the `protoc` binary.
+        ///
+        /// If this is not set, SPM will try to find the tool itself.
+        var protocPath: String?
+
+        /// A list of invocations of `protoc` with the `SwiftProtobuf` plugin.
+        var invocations: [Invocation]
+    }
+
+    static let configurationFileName = "swift-protobuf-config.json"
+
+    func createBuildCommands(context: PluginContext, target: Target) throws -> [Command] {
+        print("Creating build commands (\(UUID()))")
+
+        // Let's check that this is a source target
+        guard let target = target as? SourceModuleTarget else {
+            throw PluginError.invalidTarget
+        }
+
+        // We need to find the configuration file at the root of the target
```

Comment:
Is sourceFiles recursive?  i.e. - is this really gonna be at the root of the target or could it be found nested in a sub folder?  i.e. - which behavior do you want.

Followup - should we check to see if there are >1 files with the expected name an error since something likely isn't gonna go as the developer might have expected?

Reply:
You are right. This is actually recursive. I changed the code here to only look at the root now. AFAIK from the discussions in the forums and how our tool is working we should require it to be in the root.

I am also not diagnosing if there are multiple files. If there is no file at the root, the error looks like this
> Error Domain=NSCocoaErrorDomain Code=260 "The file “swift-protobuf-config.json” couldn’t be opened because there is no such file." UserInfo={NSFilePath=/Users/franzbusch/Developer/swift-protobuf-examples/01-same-module-usage/Sources/swift-protobuf-example/swift-protobuf-config.json, NSUnderlyingError=0x600001f3c060 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}

This Error is thrown by the `Data` init and looks sensible to me, so I wouldn't create a custom error for this